### PR TITLE
Millau -> Rialto messages relay

### PR DESCRIPTION
**Pull request by svyatonik**
~This PR should be minimal working Millau -> Rialto messages relay. Right now it won't work because #386 is still pending.~

This PR is minimal working Millau -> Rialto messages relay. Right now I'm unable to confirm if it works, because some underlying components are still missing. There are some things that needs to be fixed in follow up PRs: #432 #433 #457 . They are not critical for test deployments, that's why I have extracted them as a separate issues.

### CODE REVIEW

**Review by HCastano — COMMENTED** [review]
I'm having a bit of a hard time following the changes in this PR. There's a few things I feel like I'm missing:

- Where should I start reviewing?
- Is there's anything particularly important to look out for? 
- What's the overall approach to getting the message relay working?
- What's the expected usage?
- Etc.

You mention that there are "underlying components missing". What components are missing and why? What part of the relay doesn't work because of it? There's also a few TODOs. Which ones are going to be addressed in this PR vs. later PRs? For some TODOs its unclear

**Comment by svyatonik** [comment]
> Where should I start reviewing?
> Is there's anything particularly important to look out for?
> What's the overall approach to getting the message relay working?
> What's the expected usage?

The `message-lane` relay is originally introduced in #318 . This PR implements traits from #318, required to serve Millau -> Rialto lanes. You start relay by running `millau-messages-to-rialto` subcommand of the `substrate-relay`. The single relay instance is serving single lane.

Some implementation details of Millau -> Rialto message lane relay:
- since all the chains we're going to bridge in near future will be based on the `message-lane` + `call-dispatch` modules, the most of traits-implementation code would be the same. So the `relays/substrate/src/messages_source.rs` contains implementation of `messages_relay::SourceClient` for such chains and the `relays/substrate/src/messages_target.rs` is the implementation of `messages_relay::TargetClient`;
- everything that is specific for Millau -> Rialto lanes is in the `relays/substrate/src/millau_messages_to_rialto.rs`.

There's additional change included in this PR that is required for Millau -> Rialto message lane relay. It is in these two commits: https://github.com/paritytech/parity-bridges-common/pull/399/commits/2fe0ac355d4f57e817205209a44cff982239e27e and https://github.com/paritytech/parity-bridges-common/pull/399/commits/795d30aa407c13b1c5a8f14383572917a45ef111. Basically - we have had the same delivery strategy for both delivery and confirmation races before. It has been moved from `relays/messages-relay/src/message_race_delivery.rs` to `relays/messages-relay/src/message_race_strategy.rs`. It simply includes all pending messages/confirmations in the batch.

But this original concept needs to be changed now, because: (1) scheduled messages were removed from `message-lane` module and (2) outbound lane state proof is now required to be included into messages proof sometimes. So there's now custom implementation of the `RaceStrategy` for message delivery loop. It is the `MessageDeliveryStrategy` from the `relays/messages-relay/src/message_race_delivery.rs`.

> You mention that there are "underlying components missing". What components are missing and why?

Atm the last required component afaik is #397 . It is missing because noone works on that.

> What part of the relay doesn't work because of it?

I haven't claimed that some parts of relay doesn't work because of this :) The whole message delivery + dispatch pipeline isn't working => I can't guarantee that relay works, because all its transactions would fail.

> There's also a feew TODOs. Which ones are going to be addressed in this PR vs. later PRs? For some TODOs its unclear

Right, thanks! I have been checking the code for leftover TODOs before moving to ready-for-review state, but missed 3 of them :) Fixed in last commits. So there are now 4 TODOs left: #432 #433 #457 #468

File: relays/substrate/src/messages_source.rs
Comment (HCastano) [inline relays/substrate/src/messages_source.rs:94]: I think it would be worthwhile to typedef this

File: relays/substrate/src/messages_source.rs
Comment (HCastano) [inline relays/substrate/src/messages_source.rs:184]: Doc comment same as function name

File: relays/substrate/src/messages_source.rs
Comment (HCastano) [inline relays/substrate/src/messages_source.rs:122]: We should do something about these hardcoded Runtime API strings, we already saw how easy they were to break last week

File: relays/messages-relay/src/message_lane_loop.rs
Comment (HCastano) [inline relays/messages-relay/src/message_lane_loop.rs:117]:
```suggestion

	/// Get nonce of latest confirmed message.
```

File: relays/substrate/src/messages_source.rs
Comment (HCastano) [reply relays/substrate/src/messages_source.rs:94]: Also, why does the proof have the weight included in it?

File: relays/substrate/src/messages_source.rs
Comment (HCastano) [inline relays/substrate/src/messages_source.rs:212]: Maybe we can add a `From<(Num, Hash)>` implementation for `HeaderId` so this can just be an `.into()` call

File: relays/substrate/src/messages_source.rs
Comment (HCastano) [inline relays/substrate/src/messages_source.rs:43]: Why do we use a `PhantomData` field for `P` instead of a real field of type `P`?

**Review by HCastano — COMMENTED** [review]
I really appreciate the background and guidance on how to approach this PR. It makes reviewing the code a lot easier. That said, I didn't get through a lot of it today 😅

> You start relay by running millau-messages-to-rialto subcommand of the substrate-relay

May I suggest something like `./relay messages --from=millau --to=rialto`? Although I think this would require support for multiple networks and a bit of a refactor of the CLI UI...

> Atm the last required component afaik is #397 . It is missing because noone works on that.

If this is important to get done let us know and I'm sure someone can find time to do it

File: relays/substrate/src/messages_source.rs
Comment (svyatonik) [reply relays/substrate/src/messages_source.rs:94]: Because weight needs to be submitted with proof.

File: relays/substrate/src/messages_source.rs
Comment (svyatonik) [reply relays/substrate/src/messages_source.rs:43]: P is not meant to be instantiated

File: relays/substrate/src/messages_source.rs
Comment (svyatonik) [reply relays/substrate/src/messages_source.rs:212]: Why do you think `.into()` is better than `HeaderId()`? The `HeaderId` is here, because tuples were not intuitive - you need some context to understand what's this tuple is. Going back to `(...).into()` doesn't help imo

File: relays/substrate/src/messages_source.rs
Comment (svyatonik) [reply relays/substrate/src/messages_source.rs:94]: (I'll add a typedef, though :) )

**Comment by svyatonik** [comment]
> May I suggest something like ./relay messages --from=millau --to=rialto? Although I think this would require support for multiple networks and a bit of a refactor of the CLI UI...

I don't like it, because it requires relay to be fully symmetric (i.e. now we have Millau -> Rialto, but Rialto -> Millau is still missing). Also there'll be problem if we'll ever need to add chain-specific arguments with default values (like `maximal_balance_decrease` for balance guard from #470). You may add something like that as an additional option in future PRs.

> If this is important to get done let us know and I'm sure someone can find time to do it

I think all issues with Millau release milestone are equally important - no need to reprioritize or something :) The fact that I can't test something doesn't mean that I can't work on that.

File: relays/substrate/src/messages_source.rs
Comment (HCastano) [reply relays/substrate/src/messages_source.rs:216]: It seems a little silly to me that we're breaking up this tuple (`decoded_best_finalized_peer_on_self`) just to turn it back into a `HeaderId`

File: relays/substrate/src/messages_source.rs
Comment (HCastano) [reply relays/substrate/src/messages_source.rs]: > Because weight needs to be submitted with proof.

I was gonna ask "why" again, but it looks like you answered it with the doc comments 👍

File: relays/messages-relay/src/message_race_delivery.rs
Comment (HCastano) [inline relays/messages-relay/src/message_race_delivery.rs:280]:
```suggestion
		let confirmations_missing = latest_received_nonce_at_target.checked_sub(&latest_confirmed_nonce_at_source);
```

File: relays/messages-relay/src/message_race_delivery.rs
Comment (HCastano) [inline relays/messages-relay/src/message_race_delivery.rs:285]: `RustFmt` doesn't do a good job with string literals, but it would be easier to read if this was split on multiple lines

File: relays/messages-relay/src/message_race_delivery.rs
Comment (HCastano) [inline relays/messages-relay/src/message_race_delivery.rs:280]: Is there a problem if we have a problem if `latest_confirmed_nonce_at_source` > `latest_received_nonce_at_target`? This seems like something that shouldn't happen

**Review by HCastano — COMMENTED** [review]
I'm still working my way through the PR

File: relays/substrate/src/messages_source.rs
Comment (svyatonik) [reply relays/substrate/src/messages_source.rs:43]: `MessageLane` is actually `P` across the code :) That's because it has been `MessagePipeline` in the beginning. So this would only add more confusion imo.

File: relays/messages-relay/src/message_race_loop.rs
Comment (svyatonik) [reply relays/messages-relay/src/message_race_loop.rs:87]: I've replaced `additional_proof_required` with `ProofParameters`. It's just a protocol between delivery strategy and source client.

File: relays/messages-relay/src/message_race_loop.rs
Comment (svyatonik) [reply relays/messages-relay/src/message_race_loop.rs:125]: Irrelevant

File: relays/messages-relay/src/message_race_loop.rs
Comment (svyatonik) [reply relays/messages-relay/src/message_race_loop.rs:66]: Right now it is only `Some` in message delivery race.

File: relays/messages-relay/src/message_race_loop.rs
Comment (HCastano) [reply relays/messages-relay/src/message_race_loop.rs]: I don't think the `ProofParameters` associated type really addresses my concern here. We still have `type ProofParameters = bool`, so it only shifts the `bool` elsewhere. Imo it's still unclear as to what the `bool` should be or what it represents

**Review by HCastano — APPROVED** [review]
Looks okay, but I would get another review before merging

File: relays/messages-relay/src/message_race_loop.rs
Comment (svyatonik) [reply relays/messages-relay/src/message_race_loop.rs:66]: Then, if we'll ever add another race, this comment would need to be updated. I doubt that going through all code comments every time you change something is ok. So I'd like to avoid any race-specific comments in race-agnostic structs, functions, etc

File: relays/messages-relay/src/message_race_loop.rs
Comment (svyatonik) [reply relays/messages-relay/src/message_race_loop.rs]: Well, as I said, it is part of protocol between strategy and client. That's the only place where you actually treat `ProofParameters` as `bool`. In both places this `bool` has a name. I think it's enough

File: relays/messages-relay/src/message_race_delivery.rs
Comment (svyatonik) [reply relays/messages-relay/src/message_race_delivery.rs]: It should not happen, right. In theory :)

**Comment by HCastano** [comment]
@tomusdrw Can you review this?

File: relays/messages-relay/src/message_race_receiving.rs
Comment (tomusdrw) [inline relays/messages-relay/src/message_race_receiving.rs:170]: Same here, what's the reason for not doing `?` right after `.await`?

File: relays/substrate/src/messages_source.rs
Comment (tomusdrw) [reply relays/substrate/src/messages_source.rs:127]: IMHO it's worth to at least create constants and maybe even group in a single module, so that there is one single place we need to update in the future. This also encourages re-use if we ever run into using the same Runtime API call in different places

File: relays/substrate/src/messages_source.rs
Comment (tomusdrw) [reply relays/substrate/src/messages_source.rs:216]: I'd rather be for `HeaderId::from_tuple(...)` instead of `.into()`

**Review by tomusdrw — APPROVED** [review]
lgtm! A few grumbles left in comments, but nothing serious.

File: relays/substrate/src/messages_source.rs
Comment (svyatonik) [reply relays/substrate/src/messages_source.rs:127]: This hasn't helped last time we had a name change. Probably better would be to introduce replacement for `decl_runtime_apis! {}` macro (that would actually call `decl_runtime_apis` from within)  that will generate string constants from method names, or even auto-generate `jsonrpsee` client for these methods. Re this particular string(s) - they're going to change anyway as a part of #457 in near future - I'll extract constants there.

File: relays/messages-relay/src/message_lane_loop.rs
Comment (svyatonik) [reply relays/messages-relay/src/message_lane_loop.rs:214]: There are conditions on when to reconnect source/target. So we may need to reconnect to source, or we may need to reconnect to target, or we may need to reconnect to both. The loop isn't that large - the largest part is `log::warn` call, which we would need anyway. I don't think that introducing join + additional calls (`.left()`, `.right`, `.and_then`, `ready`) would make this loop easier to read/much smaller.

**Pull request merged**
